Split advertised address from listen address; change address detection strategy

aaronlehmann opened the pull request:
There are currently problems with "swarm init" and "swarm join" when an explicit --listen-addr flag is not provided. swarmkit defaults to finding the IP address associated with the default route, and in cloud setups this is often the wrong choice.

Introduce a notion of "advertised address", with the client flag `--advertise-addr`, and the daemon flag `--swarm-default-advertise-addr` to provide a default. The default listening address is now 0.0.0.0, but a valid advertised address must be detected or specified.

If no explicit advertised address is specified, error out if there is more than one usable candidate IP address on the system. This requires a user to explicitly choose instead of letting swarmkit make the wrong choice. For the purposes of this autodetection, we ignore certain interfaces that are unlikely to be relevant (currently `docker*`).

The user is also required to choose a listen address on `swarm init` if they specify an explicit advertise address that is a hostname or an IP address that's not local to the system. This is a requirement for overlay networking.

Also support specifying interface names to `--listen-addr`, `--advertise-addr`, and the daemon flag `--swarm-default-advertise-addr`. This will fail if the interface has multiple IP addresses (unless it has a single IPv4 address and a single IPv6 address - then we resolve the tie in favor of IPv4).

This change also exposes the node's externally-reachable address in `docker info`, as requested by #24017.

Make corresponding API and CLI docs changes.

Fixes #23828
Fixes #23877
Fixes #23995
Fixes #24587
Fixes #24017 

cc @aluzzardi @mrjana @tonistiigi @dnephin

Commit:
Split advertised address from listen address

There are currently problems with "swarm init" and "swarm join" when an
explicit --listen-addr flag is not provided. swarmkit defaults to
finding the IP address associated with the default route, and in cloud
setups this is often the wrong choice.

Introduce a notion of "advertised address", with the client flag
--advertise-addr, and the daemon flag --swarm-default-advertise-addr to
provide a default. The default listening address is now 0.0.0.0, but a
valid advertised address must be detected or specified.

If no explicit advertised address is specified, error out if there is
more than one usable candidate IP address on the system. This requires a
user to explicitly choose instead of letting swarmkit make the wrong
choice. For the purposes of this autodetection, we ignore certain
interfaces that are unlikely to be relevant (currently docker*).

The user is also required to choose a listen address on swarm init if
they specify an explicit advertise address that is a hostname or an IP
address that's not local to the system. This is a requirement for
overlay networking.

Also support specifying interface names to --listen-addr,
--advertise-addr, and the daemon flag --swarm-default-advertise-addr.
This will fail if the interface has multiple IP addresses (unless it has
a single IPv4 address and a single IPv6 address - then we resolve the
tie in favor of IPv4).

This change also exposes the node's externally-reachable address in
docker info, as requested by #24017.

Make corresponding API and CLI docs changes.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Note that this contains a minor engine-api change that will have to be opened as a PR there after this passes through design review.

aaronlehmann commented:
> Eventually, I would like to make "swarm join" determine the local IP address by asking the remote manager what IP it sees. Then only "swarm init" will need to detect or be provided an IP address. This may come as a followup PR.

I've put together a PR for this in the swarmkit repository: https://github.com/docker/swarmkit/pull/1121. Once this is finalized/merged/vendored, we can remove the strict address validation from `swarm join` that this PR adds (`--listen-addr 0.0.0.0:<port>` would always be fine).

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+	return systemAddr, nil
+}
+
+var ipv4loopback = net.IPv4(127, 0, 0, 1)
```
any way to check 127.0.0.0/8 instead? `127.1.0.1` is also a loopback address

aaronlehmann replied:
@aluzzardi: I tried this at first, but it breaks Docker-in-Docker use cases. Docker networking uses 127.0.0.0/8 addresses for `eth0` and `docker0`.

mrjana replied:
Not sure I understand. We don't use 127.0.0.0/8 for eth0 and docker0. What was breaking in dind?

aaronlehmann replied:
```
Aarons-MacBook-Pro:docker aaronl$ make shell
...
root@978ef0b89960:/go/src/github.com/docker/docker# ifconfig eth0
eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:00:02
          inet addr:172.17.0.2  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:7 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:578 (578.0 B)  TX bytes:578 (578.0 B)
```

aaronlehmann replied:
Okay, I must be dyslexic. `172` vs `127`. I guess using `IsLoopback` is right after all.

mrjana replied:
That is 172.17.x.x address

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+func resolveListenAddr(specifiedInterface, specifiedAddr string) (string, error) {
+	specifiedHost, specifiedPort, err := net.SplitHostPort(specifiedAddr)
+	if err != nil {
+		return specifiedAddr, nil
```
This should return `err`

aaronlehmann replied:
I did this on purpose but in retrospect, my reasoning doesn't make sense. I'll change it.

aluzzardi commented on api/client/swarm/init.go:
```diff
@@ -36,7 +37,8 @@ func newInitCommand(dockerCli *client.DockerCli) *cobra.Command {
 
 	flags := cmd.Flags()
-	flags.Var(&opts.listenAddr, "listen-addr", "Listen address")
+	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address")
```
What's the default?

aaronlehmann replied:
The default is `0.0.0.0:2377`, see the definition of `NodeAddrOption`.

aluzzardi commented on api/client/swarm/init.go:
```diff
@@ -36,7 +37,8 @@ func newInitCommand(dockerCli *client.DockerCli) *cobra.Command {
 	flags := cmd.Flags()
-	flags.Var(&opts.listenAddr, "listen-addr", "Listen address")
+	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address")
+	flags.StringVar(&opts.listenInterface, flagListenInterface, "", "Listen on an interface's address")
```
So if I want to use an interface I would need to pass the port in `--listen-addr` and pass the interface in `--listen-interface`, correct? (assuming `listen-addr` has a default port then I only need to pass `--listen-interface`).

If I want to use `--listen-interface` with a different port then I'd have to pass a "bogus" `--listen-addr`.

I think this is acceptable.

Alternative: `--listen-addr eth0:4242`?

aaronlehmann replied:
> If I want to use --listen-interface with a different port then I'd have to pass a "bogus" --listen-addr.

Yeah, I agree it's not ideal.

> Alternative: --listen-addr eth0:4242

I would rather not overload `--listen-addr` like this. `eth0` could be a valid hostname as well.

I thought of `--listen-interface eth0:4242`, but subinterfaces can contain `:` characters.

Any ideas?

mrjana replied:
`--listen-interface eth0/4242`?

aaronlehmann replied:
I like this. Going once, going twice...

justincormack replied:
ugh, this is really ugly. Maybe add `--listen-port`?

aaronlehmann replied:
I'm okay with adding `--listen-port` (as described in https://github.com/docker/docker/pull/24237#discussion_r69359620) if you're unhappy with putting the interface name and port in a single option. It's not my first choice because it feels a little clunky to use two options to configure what address the node binds to.

Any thoughts for or against this?

justincormack replied:
I really dont like using `/` as it is just making up a syntax. Thinking about it actually `:` is fine though, a it is only valid in the archaic alias syntax, and you should not be getting those pseudo interface names from the Go APIs I don't think. But the whole idea of specifying an interface is pretty horrible I think, seems like a case where the user should just make a script to choose the right addresses, there are too many edge cases.

cpuguy83 replied:
We already do the colon on --cluster-advertise, which accepts an ip or an interface.

aaronlehmann replied:
Let's be consistent with `--cluster-advertise`.

`--listen-addr=192.168.1.1:4567`
`--listen-addr=eth0:4567`

This happens not to need any engine-api changes.

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+		}
+		return net.JoinHostPort(systemAddr.String(), specifiedPort), nil
+	} else if specifiedInterface != "" {
+		return "", errors.New("cannot specify both --listen-interface and --listen-addr")
```
I don't quite get how this works.

We need `--listen-addr` to specify the port - how can I pass `--listen-interface` without passing a `--listen-addr`?

aaronlehmann replied:
The error message is slightly misleading. `--listen-addr 0.0.0.0:6868 --listen-interface eth0` would be a valid combination. Let's find a less confusing way to specify a non-default port.

aaronlehmann replied:
Maybe we should have three options:

```
--listen-addr
--listen-interface
--listen-port
```

`--listen-port` defaults to 2377 and is completely optional.
`--listen-addr` and `--listen-interface` are then truly mutually exclusive.

thaJeztah replied:
`--listen addr=123.45.67.8,port=1234`?

aaronlehmann replied:
@thaJeztah: I ended up going with `--listen-interface=eth0/1234`, with the port part being optional. Easier than a microformat IMHO.

thaJeztah replied:
`eth0/1234` is a microformat as well; basically, I based my suggestion on the `--mount` format, which we already have in use, and allows future expansion if needed.

aaronlehmann replied:
My concern about doing something similar to the mount format is that simply specifying an IP address or interface is less intuitive and discoverable. You have to form a key-value pair for it behind the argument even if you don't care about the port number.

I have the same concern about mounts, BTW.

thaJeztah replied:
Oh, wait we have just `--listen-address` and `--listen-interface` now of course. Yeah, I agree, sorry about the noise; I saw your suggestion for three separate flags. If we don't expect more options to be added, that is

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+func isUsableSystemIP(ip net.IP) bool {
+	return !ip.Equal(net.IPv4bcast) &&
+		!ip.IsUnspecified() &&
+		!ip.Equal(ipv4loopback) &&
```
There is actually an `ip.IsLoopback()` function that checks if it's a loopback IP (127.0.0.0/8):

https://golang.org/src/net/ip.go?s=3132:3162#L105

aaronlehmann replied:
As I mentioned in https://github.com/docker/docker/pull/24237/files/8368703fca7cfcfb9088e274cb1fbc300ac86aec#r69358757, I'm explicitly avoding this function because it breaks Docker-in-Docker use cases.

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+
+	var systemAddr net.IP
+
+	for _, intf := range interfaces {
```
Besides checking the IP (which checks for 127.*), we should also check if it's a loopback interface.

There's a flag (FlagLoopback) that tells if it's a loopback interface. One in theory could create a loopback interface with a 10.x.x.x address, this check would catch it.

aluzzardi replied:
Speaking of flags - there's also `FlagUp`. We should probably ignore interfaces that are down.

aaronlehmann replied:
Both of these are good ideas. I'll make those changes.

aluzzardi commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,159 @@
+
+func resolveInterfaceAddr(specifiedInterface string) (net.IP, error) {
+	// Use a specific interface's IP address.
+	intf, err := net.InterfaceByName(specifiedInterface)
```
We should check for FlagLoopback here

aluzzardi replied:
And FlagUp

aaronlehmann replied:
Why here? `--listen-interface lo` seems legit. In fact, the integration tests use this.

aluzzardi replied:
`--listen-interface lo` will result in the same problems we have today: broken raft and broken networking if you ever attempt to join another node to this swarm.

As I type this I realize there's no way to remote join.

I guess we should also allow --listen-addr 127.0.0.1 then?

aaronlehmann replied:
I don't understand. Why should `--listen-addr 127.0.0.1` be allowed but not `--listen-interface lo`?

mavenugo replied:
what is the resolution for this ? It seems odd that we will allow lo or 127.0.0.1. What value does it provide ?

aaronlehmann replied:
The current version of the PR allows you to specify `127.0.0.1` or `lo`. I prefer to support it because if you are setting up a swarm that will only have one node, you can use `--advertise-addr 127.0.0.1` or `--advertise-addr lo` to satisfy the condition of having an advertise address. On an embedded system, you might not actually have any other interfaces or addresses.

But I suppose I'm open to blacklisting these if people think it it's better than allowing them.

aaronlehmann replied:
(Another, similar, use case is running automated tests in an environment without a consistent networking setup)

aaronlehmann commented:
Fixed loopback address detection and added checks for interface flags in the `resolveSystemAddr` case. Still looking for a good way to reorganize the flags so specifying both an interface and a non-default port is less confusing. Any comments on https://github.com/docker/docker/pull/24237#discussion_r69359620?

aaronlehmann commented:
Implemented @mrjana's suggestion of `--listen-interface eth0/4567`.

Also went ahead and removed this detection/validation from `swarm join` in anticipation of https://github.com/docker/swarmkit/pull/1121 (I want to avoid creating a complicated dependency graph between these PRs).

aaronlehmann commented:
Are people okay with the engine-api change here (adding `ListenInterface` to `InitRequest`)? Should I go ahead and open a PR for that?

cpuguy83 commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,160 @@
+	return interfaceAddr6, nil
+}
+
+var errMultipleIPs = errors.New("could not choose a listening IP address since this system has multiple addresses - specify one with --listen-addr or --listen-interface")
```
These errors don't seem correct. This error is being generated by the daemon, no the CLI.
The errors should be pertinent to all API users and not just Docker CLI users.

We can generate an error that the client can parse and display this extra info... but I don't think having these errors in the daemon is correct.

aaronlehmann replied:
Agreed, I will revise this. I like the suggestion of parsing the error on the client side and displaying client-specific information. Is there a generally accepted way to do this, or should the client just look for a particular substring in the error?

cpuguy83 replied:
Not today, nothing like errcode or anything.
For 1.12 we are going to be sending json errors at least, but it only has a `message` field.
We'll have to check strings for 1.12, I think.

dnephin commented:
I like this solution

aluzzardi commented:
Why not smashing them together in `listen-addr`? Or call it just `--listen`?

Also, `/` rather than `:` is weird.

thaJeztah commented:
@aluzzardi reason was that `eth0` can be either a hostname or an interface, hence the idea of having two separate flags

aaronlehmann commented:
Updated to match `--cluster-advertise` behavior. You can specify either an address or interface name to `--listen-addr`. There is only one flag and there are no changes to engine-api.

cpuguy83 commented:
One thing I don't like about this is it effectively means that a simple `docker swarm init` will no longer work in most cases.

cpuguy83 commented:
I think we should choose a default... maybe lo (not likely) or the first private address we find.
Then at the very least a single node swarm will just work.

aaronlehmann commented:
> I think we should choose a default... maybe lo (not likely) or the first private address we find.
Then at the very least a single node swarm will just work.

But when you try to add a second node, it often won't work. This is the problem the PR is trying to solve. @aluzzardi and I think that this UX penalty is worth the improvement in correctness. We've had a lot of issue reports from people who don't know they may need to specify `--listen-addr`, and have trouble getting a multi-node cluster working.

Without this PR, we choose the address associated with the default route as the default. This works sometimes. In a lot of cases (like GCE), the default is wrong.

cpuguy83 commented:
@aaronlehmann What if the CLI outputs a warning (and likewise the daemon)?

aaronlehmann commented:
ping @aluzzardi WDYT?

I can understand the desire to make `docker swarm init` work without requiring much configuration. But I'm also not sure it's a good idea to guess an important value that we may get wrong, even if we warn the user that we're making a guess. That was what we're trying up to this point (minus the warning), and it didn't work out very well.

Another complication is that there's currently no way to update the address that the first manager publishes for itself in the raft log. So if the default turned out to be incorrect, you would have to throw away the whole cluster state and start over when you want to add a second manager. It would take a big, scary warning to explain this.

tonistiigi commented on api/client/swarm/init.go:
```diff
@@ -54,6 +56,9 @@ func runInit(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts initOptions
 	nodeID, err := client.SwarmInit(ctx, req)
 	if err != nil {
+		if strings.Contains(err.Error(), "could not choose a listening IP address") || strings.Contains(err.Error(), "could not find the system's IP address") {
+			return errors.New(err.Error() + " - specify one with --listen-addr or --listen-interface")
```
`--listen-interface` seems to be removed now.

aaronlehmann replied:
Fixed, thanks.

tonistiigi commented on docs/reference/commandline/swarm_init.md:
```diff
@@ -16,12 +16,13 @@ parent = "smn_cli"
+	      --force-new-cluster        Force creation of a new cluster from current state
+	      --help                     Print usage
+	      --listen-addr value        Listen address (automatically detected address with port 2377 if unspecified)
+	      --listen-interface value   Use the given network interface's address as the listen address
```
seems to be removed

aaronlehmann replied:
Fixed, thanks.

tonistiigi commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -3624,8 +3624,14 @@ Initialize a new Swarm
+  address/port combination in the form `192.168.1.1:4567`, or an interface followed by a port
+  number, like `eth0:4567`. If the port number is omitted, the default swarm listening port is
+  used.
+- **ListenInterface** – Interface name used to derive the address for inter-manager
```
One more mention of `ListenInterface`

aaronlehmann replied:
Fixed.

tonistiigi commented on integration-cli/daemon_swarm.go:
```diff
@@ -33,6 +33,9 @@ var autoAcceptPolicy = swarm.AcceptancePolicy{
 func (d *SwarmDaemon) Init(req swarm.InitRequest) error {
 	if req.ListenAddr == "" {
 		req.ListenAddr = d.listenAddr
+		if strings.HasPrefix(req.ListenAddr, "0.0.0.0") {
```
Why isn't this required for the `SwarmDaemon.Join()`

aaronlehmann replied:
Currently: For join, swarmkit ends up choosing the address associated with the default route, which works in the unit test (just like before this PR).

In the near future: https://github.com/docker/swarmkit/pull/1121 will make address detection unnecessary for joining nodes (but it depends on this PR).

I originally had `Join` doing address detection too in this PR, but I removed it in anticipation of https://github.com/docker/swarmkit/pull/1121

tonistiigi commented:
LGTM

aluzzardi commented:
@cpuguy83 We found many user reported issues were caused by an incorrect listen address.

We lose some "magic", but we avoid users potentially wasting hours trying to debug (or giving up)

cpuguy83 commented:
@aluzzardi Sure, but a warning would suffice here along with this change wouldn't it?
How is someone with docker4mac supposed to know what interface or ip address to specify when testing out swarm features... maybe d4mac has only 1 interface and it just works but I don't even know.

aluzzardi commented:
@aaronlehmann I really like the new options.

Any idea what happens though for interfaces such as `eth0:1`? Do we assume `1` to be the port?
What happens with `eth0:1:2377`?

It might be worth adding a unit testing to try those out.

aaronlehmann commented:
@aluzzardi: I tested this and golang's interface handling functions don't work with interfaces like `eth0:1`. So something like `--listen-addr eth0:1` would not work anyway. This is what convinced me to use `:` as the port separator with interfaces instead of `/`.

stevvooe commented:
A few things:

1. I am not sure I like the use of URIs for specifying these. There isn't really a standard and I am reticent of microformats, in general.
2. What if there is more than one listen address?

aaronlehmann commented:
> I am not sure I like the use of URIs for specifying these. There isn't really a standard and I am reticent of microformats, in general.

This doesn't use URIs. The syntax is `--listen-addr 192.168.1.1:2377` or `--listen-addr eth0:2377` to listen on a particular interface's IP address (which may be convenient when scripting a deployment).

> What if there is more than one listen address?

This isn't supported right now and supporting it would entail separating the concept of the advertised address from the listening address. I think this is the right thing to do, but decided with @aluzzardi not to do it in this PR, since it's somewhat orthogonal to the problem this PR is trying to solve (bad autodetection defaults).

thaJeztah commented:
What's the format for IPv6? Did we test/account for that?

aaronlehmann commented:
@thaJeztah: This uses `net.SplitHostPort`, so IPv6 will be handled properly. The format for IPv6 is `[::1]:2377`.

aluzzardi commented:
LGTM

aluzzardi commented:
@aaronlehmann Any open issue with this PR or do we just need more LGTMs?

/cc @tonistiigi @icecrime @LK4D4

aaronlehmann commented:
Ready to go as far as I'm concerned.

cpuguy83 commented:
I'm concerned about the degraded experience for simple setups, but if no one else is...

aluzzardi commented:
@cpuguy83 We can always improve that. It's a mild annoyance: if you have multiple interfaces, then you have to specify which IP to use. However, if we do magic auto detection and select the wrong one, then suddenly things start to fail. The cluster might also be in a weird shape: even though you managed to join, overlay networking (which uses the same `listen-addr`) might not be able to route packets correctly, which is tough to debug.

Also - realistically, how often do people end up with multiple interfaces?

Now that I'm thinking of it - will `docker0` and `lxcbr0` get in the way of auto detection, @aaronlehmann ?

icecrime commented:
@cpuguy83 I don't think it's that bad: auto-detection is still there, but it's just going to bail out if there's any ambiguity. I like the fact that this puts correctness first: as we improve auto-detection we may reduce the number of error cases, while never compromising on correctness.

cpuguy83 commented:
Here's the list from docker4mac:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: bond0: <BROADCAST,MULTICAST400> mtu 1500 qdisc noop qlen 1000
    link/ether da:76:b8:b6:69:48 brd ff:ff:ff:ff:ff:ff
3: dummy0: <BROADCAST,NOARP> mtu 1500 qdisc noop qlen 1000
    link/ether 1a:13:30:33:b2:0a brd ff:ff:ff:ff:ff:ff
4: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether c0:ff:ee:c0:ff:ee brd ff:ff:ff:ff:ff:ff
    inet 192.168.65.2/29 brd 192.168.65.7 scope global eth0
       valid_lft forever preferred_lft forever
5: teql0: <NOARP> mtu 1500 qdisc noop qlen 100
    link/void 
6: sit0@NONE: <NOARP> mtu 1480 qdisc noop qlen 1
    link/sit 0.0.0.0 brd 0.0.0.0
7: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop qlen 1
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
8: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop qlen 1
    link/[823] 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
9: br-126def362035: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue 
    link/ether 02:42:ab:f9:f2:fe brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-126def362035
       valid_lft forever preferred_lft forever
10: br-4b48b5de7cd4: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue 
    link/ether 02:42:b9:3d:f9:c9 brd ff:ff:ff:ff:ff:ff
    inet 172.19.0.1/16 scope global br-4b48b5de7cd4
       valid_lft forever preferred_lft forever
11: docker_gwbridge: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue 
    link/ether 02:42:53:55:87:10 brd ff:ff:ff:ff:ff:ff
    inet 172.20.0.1/16 scope global docker_gwbridge
       valid_lft forever preferred_lft forever
    inet6 fe80::42:53ff:fe55:8710/64 scope link 
       valid_lft forever preferred_lft forever
12: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue 
    link/ether 02:42:1a:80:17:83 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:1aff:fe80:1783/64 scope link 
       valid_lft forever preferred_lft forever
14: veth46d4b48@if13: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue master docker0 
    link/ether 16:dd:54:f4:fc:5c brd ff:ff:ff:ff:ff:ff
    inet6 fe80::14dd:54ff:fef4:fc5c/64 scope link 
       valid_lft forever preferred_lft forever
20: vetha5a014d@if19: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue master docker_gwbridge 
    link/ether 4e:ae:db:ca:f9:bb brd ff:ff:ff:ff:ff:ff
    inet6 fe80::4cae:dbff:feca:f9bb/64 scope link 
       valid_lft forever preferred_lft forever
104: vethfff56d5@if103: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue master docker0 
    link/ether 1a:09:6e:27:d1:88 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::1809:6eff:fe27:d188/64 scope link 
       valid_lft forever preferred_lft forever
117: br-7e848cfc6a15: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue 
    link/ether 02:42:52:da:1c:21 brd ff:ff:ff:ff:ff:ff
    inet 172.21.0.1/16 scope global br-7e848cfc6a15
       valid_lft forever preferred_lft forever
    inet6 fe80::42:52ff:feda:1c21/64 scope link 
       valid_lft forever preferred_lft forever
119: vethb61bd87@if118: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue master br-7e848cfc6a15 
    link/ether 0a:5f:f2:0a:cd:2d brd ff:ff:ff:ff:ff:ff
    inet6 fe80::85f:f2ff:fe0a:cd2d/64 scope link 
       valid_lft forever preferred_lft forever
141: veth48b5e60@if140: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue master docker0 
    link/ether 7a:8a:98:8a:3e:30 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::788a:98ff:fe8a:3e30/64 scope link 
       valid_lft forever preferred_lft forever
```

tonistiigi commented:
If docker4mac doesn't automatically detect anymore we should add "default interface" daemon argument that could be automatically set.

aaronlehmann commented:
I've added a commit which provides a `--default-listen-addr` daemon flag.

@tonistiigi @cpuguy83 @justincormack @icecrime PTAL.

aaronlehmann commented:
Changed the flag to `--default-swarm-listen-addr`.

aaronlehmann commented:
Any thoughts on `--default-swarm-listen-addr` vs `--swarm-default-listen-addr` in case we add more swarm* flags in the future?

tonistiigi commented:
@aaronlehmann That option seems too long for `TestHelpTextVerify`

I'd slightly prefer `--swarm-default-listen-addr`

aaronlehmann commented:
Changed to `--swarm-default-listen-addr` and made the help text slightly shorter.

thaJeztah commented:
looks like it needs a rebase already

aaronlehmann commented:
Rebased.

tonistiigi commented:
LGTM

thaJeztah commented on docs/reference/commandline/swarm_init.md:
```diff
@@ -84,11 +84,18 @@ The URL specifies the endpoint where signing requests should be submitted.
+networking.
+
+It is also possible to specify a network interface to listen on that interface's address;
+for example `--listen-addr eth0:2377`.
```
is `:port` optional or required?

aaronlehmann replied:
It is optional. If you don't specify it, `:2377` is implied. But I think specifying a port is good practice. Should I clarify this?

thaJeztah replied:
May be good to clarify it. Perhaps we should use the `IP[:PORT]` notation to indicate that it's optional; should we use it in the flag description as well? e.g. for `--user` we use this in the flag description;

```
-u, --user string                 Username or UID (format: <name|uid>[:<group|gid>])
```

aaronlehmann replied:
@thaJeztah: Updated.

thaJeztah commented:
err.. forgot to mention the man page; https://github.com/docker/docker/blob/master/man/dockerd.8.md

icecrime commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,163 @@
+func resolveListenAddr(specifiedAddr string, defaultListenAddr string) (string, error) {
+	specifiedHost, specifiedPort, err := net.SplitHostPort(specifiedAddr)
+	if err != nil {
+		return "", fmt.Errorf("could not parse listen address %s", specifiedAddr)
```
I'm a bit surprised by this error path: isn't it acceptable to pass an empty listen address and rely on the default?

tonistiigi replied:
To the API `listenAddress` is a required field. We have validation for it in https://github.com/docker/docker/blob/07dd69df8d6c5f4da0059d49e128fef9277fcb5d/daemon/cluster/cluster.go#L1173

aaronlehmann replied:
@icecrime: The issue is that the port is chosen client-side. So currently `0.0.0.0:2377` is a valid value for `ListenAddr`, but an empty string is not.

We could potentially move the default port selection to the daemon side, but this has UX impacts. From the client side, the default of `--listen-addr` would change to an empty value, so it would not be as obvious what port number gets used if you don't specify anything. We would probably want to make `--swarm-default-listen-addr` take an optional port value, and then we would have to handle corner cases like no port number being specified in either argument.

Ultimately I think the choice of port belongs on the client side, but I'm curious what you think.

tonistiigi replied:
@aaronlehmann We have default port adding in daemon side as well. https://github.com/docker/docker/blob/07dd69df8d6c5f4da0059d49e128fef9277fcb5d/daemon/cluster/cluster.go#L1175

aaronlehmann replied:
@tonistiigi: If I'm reading this right, it means that the error path here should never be reached, correct? (Because `validateAndSanitizeInitRequest` adds the default port if there is no port).

Are you happy with the current approach, or should we change it to accept an empty `ListenAddr`, which would be handled the same way as `0.0.0.0`?

tonistiigi replied:
I'm fine with not accepting an empty string in API. In API level this can easily mean that user just forgot that this field existed and will likely have incorrect behavior then. Otoh port/host completion is nice because if you are used to specifying port or host only in cli you would not know that these are not supported in API level.

aaronlehmann commented:
@thaJeztah: Updated the manual page. Does the formatting look correct?

thaJeztah commented on man/dockerd.8.md:
```diff
@@ -239,6 +240,10 @@ output otherwise.
 
+**--swarm-default-listen-addr**=*IP|HOSTNAME|INTERFACE*
+  Set default address or interface for swarm listening address. This can be a hostname, an IP address, or
+  or an interface such as `eth0`. A port cannot be specified through with this option.
```
s/through/though/ or remove

thaJeztah replied:
Also there's one "or" too many

aaronlehmann replied:
Thanks, fixed the description.

aaronlehmann commented:
An update on next steps: The main thing missing from this PR is the separation of "listen address" from "advertised address". Once we introduce "advertised address", that will be the main knob users have to tweak, since binding to 0.0.0.0:2377 or [::]:2377 will usually be fine. The daemon flag will change to `--swarm-default-advertise-address`, since it's not as useful to set a default for the actual bind address.

I've started working on this. It will take some changes to engine-api and swarmkit. I feel it would be easiest to review if this PR is merged first and then the advertise-addr split is done in a separate PR. But I've heard some feedback that it's better to make all the related changes at once. So let me know...

aaronlehmann commented:
Here is a commit to create a separate `advertise-addr` flag and switch the daemon flag to providing the default for `advertise-addr`: https://github.com/aaronlehmann/docker/commit/cbb06ecf578343b96aa50a4394cc6c38587b8747

Please let me know if you'd prefer me to include it in this PR, or wait until this PR is merged to open it as a separate one.

Note that it involves upstream changes to libnetwork, swarmkit, and engine-api.

thaJeztah commented:
We were just discussing this, and currently, Swarm mode does *not* support dynamic IP-addresses for managers. If we support `eth0:4567` as a notation, this gives the impression that dynamic ip-addresses are supported ("if my IP-address changes, docker will pick up the new address from `eth0`").

For that reason, I think we should **not** support the `interface:port` notation for now, and require the user to specify the IP-address. We can add back the `interface:port` notation once we support dynamic IP-addresses.

cpuguy83 commented:
@thaJeztah I think this is fine for the daemon flag, though (and probably even a requirement).

aaronlehmann commented:
@thaJeztah: Are you okay with removing support for interface names from `--listen-addr` but keeping that support in the daemon flag?

Also, what do you think about incorporating the changes in https://github.com/aaronlehmann/docker/commit/cbb06ecf578343b96aa50a4394cc6c38587b8747 into this PR vs. having those be a followup? In that commit, there's `--listen-addr` and `--advertise-addr` on the client side. Should neither support an interface name?

thaJeztah commented:
@aaronlehmann my primary concern is above, so any setting that "allows" being dynamic could use the interface name. I think there's two issues; the "impression" that we allow dynamic IP-addresses, and possibly an actual issue (what happens if I restart a daemon/node and `eth0` got a new IP-address; does the daemon pick up the new IP address? Does the user get notified?).

IIUC, having "interface" may be required for Docker for Mac to determine the right address; especially in that case (restarting laptop / switching WiFi) may be troublesome, so I'm not sure what's best here.

I may not be the best person to ask, and happy to get more input (don't want to be a PITA, just some concerns)

aaronlehmann commented:
> IIUC, having "interface" may be required for Docker for Mac to determine the right address; especially in that case (restarting laptop / switching WiFi) may be troublesome, so I'm not sure what's best here.

I don't think this is particularly important for Docker for Mac, because the interfaces in the Linux environment are the ones that matter, and I don't think they would change if the Mac switches networks.

@aluzzardi @justincormack: Is there any hard requirement to support using interface names as values to these flags? Would it be okay to defer this until we support dynamic addressing for managers (probably post 1.12)? I guess the motivation for supporting this was that you could easily write scripts that use the correct interface without going through the steps to determine the actual address. @thaJeztah raises the concern that this approach might make it look like we'll adjust properly to an address change. I'm not sure what the best solution is, because I see his point, but also think it would be awkward to need to pass the address directly if you want to write a generic script (though https://github.com/docker/swarmkit/pull/1121 will make it so you only need to do this on the first node, so maybe it's not as big a deal).

Let's find an acceptable solution so we can move this forward.

cpuguy83 commented:
@thaJeztah The daemon flag, though, is just setting a default value. Once it has joined a swarm, the user should not expect that changing a default value would changing something that's already set.

aluzzardi commented:
Hey, sorry, didn't look at the diff yet but did we manage to make detection work on most typical environments?

aaronlehmann commented:
Detection works in environments with one IP. When there are multiple IPs, you can specify it explicitly or use a daemon flag that specifies the default IP or interface.

aluzzardi commented:
@aaronlehmann But did we change something regarding bridge interfaces?

As I pointed out last time, while doing manual research, I didn't find a single environment that didn't include `docker0`, `docker_gwbridge`, `lxcbr0` or some other kind of bridge interfaces out of the box.

I believe merging this is as good as making `--listen-addr` mandatory since I couldn't find a "real world machine" (as opposed to dind / dropping binaries) that actually only had a single interface once docker was installed.

aaronlehmann commented:
@aluzzardi: We could whitelist specific interface names that we don't think would be reasonable defaults. Other than that, I'm not sure what to suggest.

I think that this is a UX optimization that could easily come afterwards. Let's focus on getting something in place that's not broken. As I write this, there are 113 comments on this PR suggesting many different tweaks. It feels like it's only getting harder to get consensus around how this should work. Meanwhile, users trying the rc releases are finding that setting up multinode clusters doesn't work, and they keep submitting bugs. I have 4 other PRs to improve usability around joining nodes to a swarm cluster that are blocked on this one.

So while I welcome the input, I'd really like to get this PR unblocked instead of adding more magic at this point. I'm afraid that if I go and add a whitelist like this, we're going to spend weeks discussing exactly what should be on that whitelist.

At the moment this PR is blocked by concerns over specifying interface names to the flags (see https://github.com/docker/docker/pull/24237#issuecomment-231495827 and the few comments right after it). I'd definitely appreciate any input you have on this.

justincormack commented:
If we don't have advertise address it does not LGTM.

This cannot be split, as listen address will be 0.0.0.0 in almost every
case once we have advertise address.
On 13 Jul 2016 10:05 p.m., "Andrea Luzzardi" <notifications@github.com>
wrote:

@aaronlehmann <https://github.com/aaronlehmann> But did we change something
regarding bridge interfaces?

As I pointed out last time, while doing manual research, I didn't find a
single environment that didn't include docker0, docker_gwbridge, lxcbr0 or
some other kind of bridge interfaces out of the box.

I believe merging this is as good as making --listen-addr mandatory since I
couldn't find a "real world machine" (as opposed to dind / dropping
binaries) that actually only had a single interface once docker was
installed.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/pull/24237#issuecomment-232486471>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAdcPGkdCfl3jkLPAuGK0RHH0fuqjnyLks5qVVMcgaJpZM4JDc-x>
.

aaronlehmann commented:
@justincormack: Can you please take a look at https://github.com/aaronlehmann/docker/commit/cbb06ecf578343b96aa50a4394cc6c38587b8747, which adds advertise addresses? I understand your comment to mean you'd prefer including that as part of this PR, which is okay with me.

rneugeba commented:
A couple of general comments (but I have to admit that I don't fully understand the issue this PR is trying to solve):
- I think introducing a listening interface like `eth0:2377` is a bad idea. It's an L3/L4 service and the interface is L1/L2. Not only can the interface change (as noted by others) but the routing can change too.
- One could listen on `0.0.0.0` and then use `SO_BINDTODEVICE` but that's not portable (e.g. it doesn't work on Windows).

There were some questions regarding Docker for Mac/Windows:
- Docker4Mac uses a single interface (`eth0`) for external connectivity and its IP address is fixed, but not externally exposed, ie its purely used for host<->VM communication and the ports from the VM are exposed on the host where it is listening on `0.0.0.0`.
- Docker4Windows actually uses *two* interfaces (`eth0` and `eth1`) with `eth1` being the default gateway based on routing metrics alone (there are two default gateway entries). `eth1` "can" be configured the same as `eth0` on the Mac (and is now as default) with ports being exposed on localhost on the host, but there is still an option and possibility for traffic from the host to go via `eth0`. The Docker4Windows networking set-up is still subject to change, though.

justincormack commented:
@aaronlehmann yes https://github.com/aaronlehmann/docker/commit/cbb06ecf578343b96aa50a4394cc6c38587b8747 looks good to me, definitely what we want not this version.

aluzzardi commented:
> I think that this is a UX optimization that could easily come afterwards.

@aaronlehmann LGTM to unblock but we have to acknowledge the following:

This PR **removes** support for auto detection. There is some code that a handful of Docker maintainers will hit when performing some dind testing but for everyone else, `--listen-addr` is now **mandatory**. 

We can decide now whether we're going to fix it for 1.12 or remove auto detection as a feature from our docs.

aaronlehmann commented:
I've updated this PR to include the "listen address"/"advertised address" split. This is a more involved change than the original version of the PR, but the upside is that it's a more complete form that can be reviewed without thinking ahead to future changes.

Note that this involves changes to engine-api, libnetwork, and swarmkit, which will have to be merged upstream before this PR is merged. The swarmkit PR is https://github.com/docker/swarmkit/pull/1150. I'll open the other two once this PR gets back into code review.

aluzzardi commented:
@aaronlehmann @mavenugo I didn't follow the earlier discussion about auto detecting bridges (netlink?). Was there a fundamental problem with that? Or could we do a hardcoded blacklist instead? Not saying we should do this in this PR, but we must at least log an issue for either 1.12 or 1.13.

aaronlehmann commented:
It feels wrong to me to treat bridges differently than any other kind of interface. I think we would end up with corner cases where running without arguments silently does the wrong thing, which is what we're trying to avoid. If the goal is to avoid getting confused by docker-specific bridges, I would feel more comfortable specifically whitelisting or blacklisting (depending how you think about it) those interface names.

aaronlehmann commented:
@aluzzardi What if we ignore interfaces named docker* to begin with? We can come back and add others as necessary, but I think that might address most of your concern.

We can add others if necessary.

Aaron

> On Jul 14, 2016, at 16:01, Andrea Luzzardi <notifications@github.com> wrote:
> 
> @aaronlehmann @mavenugo I didn't follow the earlier discussion about auto detecting bridges (netlink?). Was there a fundamental problem with that? Or could we do a hardcoded blacklist instead? Not saying we should do this in this PR, but we must at least log an issue for either 1.12 or 1.13.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

aaronlehmann commented:
Added two things:

- Ignore certain interfaces when determining the system address. Currently this ignores `docker*`. @aluzzardi: Does this address your concern?
- Require a listen address to be specify when the advertise address is not one of the system addresses. This is needed for libnetwork.

aluzzardi commented:
@aaronlehmann We should probably also blacklist `lxcbr0` (or `lxcbr*`). That interface shows up in some vanilla Docker installations from what I can see.

@mavenugo Any other interface we should blacklist for auto detection? `docker_gwbridge` is already being taken care of by `docker*`.

What about our `veth*` pairs or `ovs*` interfaces? Would there by a way to make sure all of docker's interfaces are prefixed by `docker`?

mavenugo commented:
@aluzzardi @aaronlehmann I think we should blacklist all the interfaces with an ip-address that belongs to any of the `docker network` subnet. That will automatically remove all the docker created interfaces and bridges.

tonistiigi commented on daemon/cluster/cluster.go:
```diff
@@ -239,8 +293,16 @@ func (c *Cluster) startNewNode(forceNewCluster bool, listenAddr, joinAddr, secre
+
+	// Do this assignment after saveState so that the result of
+	// the resolution won't be saved persistently.
+	c.localAddr = actualLocalAddr
```
On reconnection it seems to be reusing `c.localAddr` and still saving it. I'm not actually sure about saving the `remoteAddr` on disk and reusing it. It is likely only valid on the time of the request. If I take down that node that was used as a remote for a specific node I don't want it to affect that node. If I understand it correct this case may even mean that my node doesn't start anymore.

aaronlehmann replied:
> On reconnection it seems to be reusing c.localAddr and still saving it.

I think this case needs to be fixed. I guess I will split the fields in `Cluster` between the local addr that was passed into `startNewNode`, and the one we are actually using.

> I'm not actually sure about saving the remoteAddr on disk and reusing it. It is likely only valid on the time of the request. If I take down that node that was used as a remote for a specific node I don't want it to affect that node.

`remoteAddr` is only ever used to find `localAddr` in the case where `localAddr` is not given. Basically it's used with a UDP socket to query the kernel's routing table and find what local IP we would use when connecting to that node. It doesn't matter if that node still exists, and we won't affect that node in any way.

I think saving `remoteAddr` so that we can late-bind `localAddr` makes sense, but let me know if you prefer a different way.

aaronlehmann replied:
> On reconnection it seems to be reusing c.localAddr and still saving it.

I've updated the PR to fix this.

I also squashed the commits and implemented @mavenugo's suggestion of filtering all network interfaces that contains an address in any `docker network` subnet.

aaronlehmann commented:
Implemented @mavenugo's suggestion to ignore interfaces whose IP addresses fall into subnets of `docker networks` networks. Seems to work well. Also rebased and squashed the commits.

tonistiigi commented on daemon/cluster/cluster.go:
```diff
@@ -74,28 +76,52 @@ var defaultSpec = types.Spec{
+// NetControllerProvider exposes a function for retrieving the network
+// controller.
+type NetControllerProvider interface {
+	NetController() libnetwork.NetworkController
```
Would it make sense to define a custom method for the specific data you need to detect the interfaces. `Cluster` doesn't  need a `NetworkController` and we want to avoid someone accidentally using this to mess with local libnetwork state.

aaronlehmann replied:
I could make a function that returns the subnets of networks that Docker manages. Do you think that would be better?

tonistiigi replied:
Yes. Thanks.

aaronlehmann replied:
@tonistiigi: Done.

aaronlehmann commented:
Opened engine-api PR here: https://github.com/docker/engine-api/pull/326

This can be merged in advance of this PR.

The swarmkit PRs are:

https://github.com/docker/swarmkit/pull/1119
https://github.com/docker/swarmkit/pull/1121
https://github.com/docker/swarmkit/pull/1150

Those need to be coordinated with the merging of this PR.

mavenugo commented:
@aaronlehmann am trying the latest changes and I see that the `GetLocalAddress` is empty for the following case (dind) :

```
$  sudo docker run --name=docker1 -id --privileged  -p 1234:1234 -v "/home/madhu/workspace-go/src/github.com/docker/docker/bundles:/go/src/github.com/docker/docker/bundles" -v /home/madhu/workspace-go/src/github.com/docker/docker/bundles/latest/binary-daemon/docker-containerd:/usr/bin/docker-containerd -v /home/madhu/workspace-go/src/github.com/docker/docker/bundles/latest/binary-daemon/docker-runc:/usr/bin/docker-runc -v /home/madhu/workspace-go/src/github.com/docker/docker/bundles/latest/binary-daemon/docker-containerd-shim:/usr/bin/docker-containerd-shim -v /home/madhu/workspace-go/src/github.com/docker/docker/bundles/latest/binary-daemon/docker-swarmd:/usr/bin/docker-swarmd "docker-dev:aaronl-listen-addr-revisions" ./bundles/latest/binary-daemon/dockerd  -H tcp://0.0.0.0:1234 -D
95523517d766524d6649432447a53d18d8439e9925940a47dcad4e60d6014b46

root@Ubuntu-vm docker (aaronl-listen-addr) $  sudo DOCKER_HOST=tcp://172.17.0.1:1234 ./bundles/latest/binary-client/docker swarm init --secret="" --auto-accept manager --auto-accept worker --listen-addr $(sudo docker inspect --format '{{.NetworkSettings.IPAddress}}' docker1):4500
Swarm initialized: current node (aon4arinl4t3yij27cws5ugro) is now a manager.

To add a worker to this swarm, run the following command:
    docker swarm join  \
    --ca-hash sha256:ee8615607be22edb3b53893572eea64eb5213b70b4b2bd704da7a524206a7291 \
    172.17.0.2:4500
```

I was hoping to see GetLocalAddress never to return empty string.
BTW, GetAdvertiseAddress returns 172.17.0.2.

I will review the code to spot the issue. (BTW, I didnt see this issue in the previous iterations of this PR).

mavenugo commented on api/client/swarm/init.go:
```diff
@@ -43,7 +47,8 @@ func newInitCommand(dockerCli *client.DockerCli) *cobra.Command {
 	flags := cmd.Flags()
-	flags.Var(&opts.listenAddr, "listen-addr", "Listen address")
+	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address (format: <ip|hostname|interface>[:port])")
+	flags.StringVar(&opts.advertiseAddr, flagAdvertiseAddr, "", "Advertised address (format: <ip|hostname|interface>[:port])")
```
As discussed, shall we support just the case of ip-address ? 
we can support the hostname in subsequent releases ?

abronan replied:
@aaronlehmann As using hostnames is not supported for now. Let's remove it from the flag comment to avoid any confusion.

aaronlehmann replied:
Whoops, updated it in the doc but missed the acutal flag definitions. Fixing.

Commit:
Require listen address and advertise address to be an IP address or an interface name

Hostnames are not supported for now because libnetwork can't use them
for overlay networking yet.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
@mavenugo: I've added a commit that restricts the listen address and advertise address to an IP address or interface name (cc @abronan).

I also found and fixed the bug where `GetLocalAddress` could return an empty string. It was really stupid.

abronan commented on api/client/swarm/join.go:
```diff
@@ -35,6 +37,7 @@ func newJoinCommand(dockerCli *client.DockerCli) *cobra.Command {
 
 	flags := cmd.Flags()
 	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address")
+	flags.StringVar(&opts.advertiseAddr, flagAdvertiseAddr, "", "Advertised address (format: <ip|hostname|interface>[:port])")
```
same comment here

Commit:
Vendor libnetwork and github.com/vishvananda/netlink

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

abronan commented on hack/vendor.sh:
```diff
@@ -60,12 +60,12 @@ clone git golang.org/x/net 2beffdc2e92c8a3027590f898fe88f69af48a3f8 https://gith
 
 #get libnetwork packages
-clone git github.com/docker/libnetwork 905d374c096ca1f3a9b75529e52518b7540179f3
+clone git github.com/docker/libnetwork 9d9aa530c445e2274b8044523e3c0a8cdf56f0af https://github.com/mavenugo/libnetwork
```
We can switch back and use the main repository here instead of the fork

abronan commented on man/dockerd.8.md:
```diff
@@ -55,6 +55,7 @@ dockerd - Enable daemon mode
 [**-s**|**--storage-driver**[=*STORAGE-DRIVER*]]
 [**--selinux-enabled**]
 [**--storage-opt**[=*[]*]]
+[**--swarm-default-advertise-addr**[=*IP|HOSTNAME|INTERFACE*]]
```
remove `Hostname`

abronan commented on man/dockerd.8.md:
```diff
@@ -239,6 +240,11 @@ output otherwise.
 **--storage-opt**=[]
   Set storage driver options. See STORAGE DRIVER OPTIONS.
 
+**--swarm-default-advertise-addr**=*IP|HOSTNAME|INTERFACE*
```
same comment here

abronan commented:
Overall LGTM

Have been testing manually with the address detection on the Swarm side and everything fits nicely. We should make sure we can use `hostnames` in the future though.

aaronlehmann commented:
@abronan: Addressed your feedback, thanks.

This PR is now basically ready to go. The only thing that still needs to happen is a proper revendoring of swarmkit. Swarmkit doesn't have the upstream PRs merged yet because they need to be merged at the same time as this PR. Once I get LGTMs, I'll merge those Swarmkit PRs and fix the vendoring.

(Note swarmkit isn't directly vendorable in its current state because the swarmkit PR for join tokens has already gone in, but we can work around that if we merge this one before #24823.)

mavenugo committed:
Daemon changes to pass exec-root to libnetwork

This is required to make the libnetwork's namespace mgmt
directory configurable

Signed-off-by: Madhu Venugopal <madhu@docker.com>

Commit:
Revendor swarmkit

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Rebased

mavenugo commented on api/client/system/info.go:
```diff
@@ -86,6 +86,7 @@ func runInfo(dockerCli *client.DockerCli) error {
 			fmt.Fprintf(dockerCli.Out(), " Managers: %d\n", info.Swarm.Managers)
 			fmt.Fprintf(dockerCli.Out(), " Nodes: %d\n", info.Swarm.Nodes)
 		}
+		fmt.Fprintf(dockerCli.Out(), " Node Address: %s\n", info.Swarm.NodeAddr)
```
@aaronlehmann shouldnt we expose both the LocalAddress (which is supposed to be the node address) and also the advertise-address ? I think that will be useful for cases such as using public-ip for advertise.

aaronlehmann replied:
@mavenugo: It's exposing the advertise address here. Do you think it should expose the local address as well? Is there a use case for getting that information?

mavenugo replied:
I was thinking more for debugging. We can always add it if there is a need later.

mavenugo commented on daemon/cluster/listen_addr.go:
```diff
@@ -0,0 +1,268 @@
+			return "", "", errBadDefaultAdvertiseAddr
+		}
+
+		return c.config.DefaultAdvertiseAddr, listenAddrPort, nil
```
This `if` block and the previous one are almost identical and potentially be optimized by moving it in a function and calling it twice. But it is not a blocker.

mavenugo commented on hack/vendor.sh:
```diff
@@ -139,7 +139,7 @@ clone git github.com/docker/docker-credential-helpers v0.3.0
 
 # cluster
-clone git github.com/docker/swarmkit 38857c06dafcf939a56d2650d8e0011b5aace384
+clone git github.com/docker/swarmkit 0aaad148d8c0498f1bd39d7471c9ed0e585620c9 https://github.com/aaronlehmann/swarmkit
```
should this be fixed ?

aaronlehmann replied:
Yes. I am waiting for LGTMs before I merge the upstream PRs in swarmkit. They will block other PRs from vendoring swarmkit until this one is merged, so I am waiting until the last minute.

mavenugo commented:
@aaronlehmann only a few minor (non-blocking) comments. 
I tested a few cases and they seem to be working as expected. I will continue to try more cases.

LGTM

tiborvass commented:
LGTM

mavenugo commented:
@aaronlehmann the failure in experimental CI is consistent `(11:49:41 FAIL: docker_api_swarm_test.go:728: DockerSwarmSuite.TestApiSwarmForceNewCluster)`

Its not clear why Janky didnt catch the same, but failing consistently in experimental. Can you PTAL ?

aaronlehmann commented:
Yes, I've been looking into it. I think it's unrelated to this PR, but seems to be caused by a commit in the vendored swarmkit.

mavenugo commented on integration-cli/check_test.go:
```diff
@@ -211,7 +211,7 @@ func (s *DockerSwarmSuite) AddDaemon(c *check.C, joinSwarm, manager bool) *Swarm
 	}
 	d.listenAddr = fmt.Sprintf("0.0.0.0:%d", d.port)
-	err := d.StartWithBusybox("--iptables=false") // avoid networking conflicts
+	err := d.StartWithBusybox("--iptables=false", "--swarm-default-advertise-addr=lo") // avoid networking conflicts
```
Won't this change to use `loopback` will result in `TestApiSwarmForceNewCluster` to fail ? , especially because of the way `AddDaemon` is constructed. Both the daemons will advertise 127.0.0.1 as the address to reach and the 2nd daemon (manager) tries to join the first daemon, it will try to connect to its own loopback address ?

mavenugo replied:
Also this rolls into my previous question on supporting `loopback` as an option.

aaronlehmann replied:
This is not the problem. It automatically adds the correct port, and there's logic in `AddDaemon` to allocate ports for the different daemons.

`TestApiSwarmForceNewCluster` only fails sometimes. I am also seeing `TestApiSwarmServicesUpdate` and `TestApiSwarmServicesMultipleAgents` fail intermittently.

I'm running into the same failures in #24943, so it does not seem to be related to the code in this PR. It looks like a problematic commit in the vendored swarmkit. I just eliminated one commit as a possibility and I'll keep trying others. It's tricky because I can't reliably reproduce any of the failures locally.

mavenugo commented:
@aaronlehmann could you pls resolve the conflicts ? Thanks for resolving the IT issue. This PR is important and blocking other upcoming vendor-in patches. lets get this in soon after the conflicts are resolved.

aaronlehmann commented:
All green except Windows CI, which appears not to be operational:

```
16:23:58 java.io.IOException: remote file operation failed: \gopath\src\github.com\docker\docker at hudson.remoting.Channel@1b85c535:azure-windows-rs1-5: java.nio.file.FileSystemException: \gopath\src\github.com\docker\docker\integration-cli: The process cannot access the file because it is being used by another process.
```

Merging.
